mllib/src/main/scala/org/apache/spark/ml/tree/Node.scala

```diff
@@ -287,6 +292,41 @@ private[tree] class LearningNode(
     }
   }
 
+  /**
+   * Method testing whether a node is a leaf.
+   * @return true iff a node is a leaf.
+   */
+  private def isLeafNode(): Boolean = leftChild.isEmpty && rightChild.isEmpty
+
+  /** True iff the node should be a leaf. */
+  private lazy val shouldBeLeaf: Boolean = leafPredictions.size == 1
+
+  /**
+   * Returns the set of (leaf) predictions appearing in the subtree rooted at the considered node.
+   * @return the set of (leaf) predictions appearing in the subtree rooted at the given node.
+   */
+  private def leafPredictions: Set[Double] = {
+
+    val predBuffer = new scala.collection.mutable.HashSet[Double]
+
+    // collect the (leaf) predictions in the left subtree, if any
+    if (leftChild.isDefined) {
```

Comment:
You can write things like `leftChild.foreach(predBuffer ++= _.leafPredictions)`,  or something about like that

Reply:
I originally wanted to write something similar, but given that there is not "getChildren" method, I should put leftChild and rightChild in a collection, then filter on _.isDefined, and finally go for the foreach. It globally seemed cleaner the version proposed in the PR. What do you think?

Reply:
No, I'm just thinking you use `Option.foreach(...)` twice. It's a matter of taste I guess. I tend to prefer avoiding the `isDefined` and `get` calls for simple one-liners.

Reply:
Ok, now I get what you mean. In this case though, we keep the if to apply "short-circuiting" as suggested below.